### PR TITLE
[Enhancement] Introduce Link Target in Image Block

### LINKED ISSUE #9458: Option to open an image in a new window missing from link options
Create an image block.

Tell WordPress to assign a link to the image via "Link Settings".

No option to open the link in a new window.

The option is present in text links in other blocks.

This feature is present in Wordpress 4, not present in Gutenberg. Assuming it is a bug / omission.

### PR DESCRIPTION
**Pull request by nfmohit**
## Description
This PR closes #9458 which requests the addition of an option to open the image in a new tab under "Link Options" in the Image block.

## How has this been tested?
This PR has been tested by going through the following steps:

1. Started a new post using the Gutenberg editor.
2. Added the "Image" block.
3. Made sure the "Open in New Window" option shows up under "Link Options".
4. Made sure the link target functionality works in the front-end.

## Screenshots 
pull-9458-min
(Apologies about the crappy screencast, had to shrink the browser down and compress for Github to accept it).

## Types of changes
This PR introduces a new `boolean` attribute named `linkTarget`, which if `true`, applies the `_blank` target attribute. It adds `ToggleControl` in the `InspectorControls`, which toggles the attribute `onChange`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by ayersc3** [comment]
I've seen this option for opening links in new windows for some, but not all. I've noticed it available when linking the image to the media file. But, not available when linking to a custom URL. ...Thoughts?

**Comment by cristian-e** [comment]
Tested the code and works fine. Just a suggestion: you can replace the `onChange` attribute of `ToggleControl` to:
```js
onChange={ () => setAttributes( {  linkTarget: ! linkTarget } ) }
```
This way you no longer need the `toggleAttribute` function.

**Comment by ayersc3** [comment]
Sorry for the delayed response, and thanks for your responses!

@nfmohit-wpmudev : 
WP v4.9.8
Browser: Google Chrome (v68.0.3440.106)
OS: Mac (v10.13.6)
Guttenberg: v3.7 (As I'm writing this, I upgraded to the v3.8, but still was missing the option to "open in new window")



@cristian-e 
Thanks for the code suggestion. I'll try that today!

screen shot 2018-09-17 at 9 39 45 am

**Comment by nfmohit-wpmudev** [comment]
Thank you so much for the tip @cristian-e ❤️ I've included it in the latest edited commit.

About the issue you are seeing @ayersc3, I'm extremely sorry but I wasn't able to replicate it by testing it in different environments. Just to confirm, may I know how you are testing it? Are you fetching the PR branch or just copying the code? May I know if you are re-building the `npm` project after the PR is merged locally? Thank you for your help ❤️

**Comment by chrisvanpatten** [comment]
Could we get a design review on this and potentially land it in 4.1? This and #10128 feel like easy wins for users.

**Review by talldan — COMMENTED** [review]
Thanks for another awesome contribution @nfmohit-wpmudev 🔥🔥🔥 

I've left a couple of review comments.

**Comment by nfmohit-wpmudev** [comment]
Implemented. Thank you so much for the review @talldan ❤️ Couldn't make it without your help in DM regarding the toggle value calculation logic ❤️

File: packages/block-library/src/image/edit.js
Comment (afercia) [inline packages/block-library/src/image/edit.js]: Worth considering in core all occurrences of `(opens in a new window)` have been changed to `(opens in a new tab)`, see https://core.trac.wordpress.org/ticket/43803 which will be released in 5.1 (not 5.0). Even if that string is meant for visually hidden accessible text, for consistency I'd suggest to refer to "tab" and not "window".

File: packages/block-library/src/image/index.js
Comment (nfmohit-wpmudev) [reply packages/block-library/src/image/index.js]: Implemented in [this commit](https://github.com/WordPress/gutenberg/pull/9520/commits/c49d907e664ce6c2a0e7da8af05ee8a3fd69e8b9).

File: packages/block-library/src/image/index.js
Comment (nfmohit-wpmudev) [reply packages/block-library/src/image/index.js]: Implemented in [this commit](https://github.com/WordPress/gutenberg/pull/9520/commits/c49d907e664ce6c2a0e7da8af05ee8a3fd69e8b9).

**Comment by talldan** [comment]
@nfmohit-wpmudev This is looking really close. I just spotted one thing that I missed earlier, 'target' will need to be added to this array:
https://github.com/WordPress/gutenberg/blob/aead14ac3fd919a2fdeb510b3d492d368e506f2d/packages/block-library/src/image/index.js#L89

That ensures the attribute is parsed and set correctly. To test it you can try the following steps:
1. Add an image block and upload an image (or select one from the media library).
2. Choose an option under 'Link to' and set the link to open in a new tab.
3. From the block menu, switch to Edit as HTML
4. Copy the HTML and paste it into a new paragraph
5. The pasted content should be transformed into a new block.
6. Check the sidebar 'Opens in a new tab' setting, it should be correctly enabled to reflect that the pasted html had the `target="_blank"` attribute.

Without the 'target' value in the array, you'll see that 'Opens In a New Tab' remains off, and when inspecting the HTML, target is stripped off.

**Comment by talldan** [comment]
It might also be good to rebase this against master - there was at least one conflict when I tested (pretty easy to resolve though). Interesting though that it's not showing in the github UI as a conflict. 🤔

**Comment by nfmohit-wpmudev** [comment]
@talldan Thank you so much for pointing out the array @talldan ❤️ It was completely out of my consideration. I've addressed that and the conflicts with master in the latest edited commit.

**Review by talldan — APPROVED** [review]
This is looking good to me now. Thanks for addressing all the changes so quickly, @nfmohit-wpmudev ❤️ 

We should hold off merging until the 4.1 milestone has been shipped, but once that's done this is good to go.

**Comment by youknowriad** [comment]
I'd appreciate a design validation as well :)

File: packages/block-library/src/image/index.js
Comment (aduth) [inline packages/block-library/src/image/index.js]: Does `rel` need to be here?

Comment (nfmohit-wpmudev) [reply packages/block-library/src/image/index.js]: @aduth I'm honestly not completely sure but according to [this comment](https://github.com/WordPress/gutenberg/pull/9520#issuecomment-430961765), it seems the `rel` value gets set correctly when the HTML is copied over to a new paragraph element even without `rel` attribute added to the array. @talldan Any comments here?

**Review by karmatosed — APPROVED** [review]
I am approving as can see it being useful but have to say that sidebar is getting really packed with options for the image block. I would encouraging future thinking about how we can reduce the options and iterate.

File: packages/block-library/src/image/index.js
Comment (talldan) [reply packages/block-library/src/image/index.js]: At the moment `rel="noreferrer noopener"` is added to any link with `target="_blank"` by the save function, which I think is a sensible default. I suppose some users might want a finer level of control over the attribute.

I think this is probably ok as a start, the code doesn't preclude more handling of `rel` being added later.

Comment (aduth) [reply packages/block-library/src/image/index.js]: It's not clear to me as someone implementing a block with a schema why the parser would disregard `rel`. I'm left confused why I should or shouldn't include the attribute here.

Comment (talldan) [reply packages/block-library/src/image/index.js]: It doesn't seem to be something mentioned in the handbook (nothing about raw transformations in general, as far as I could see), so potentially a gap in the information we provide.

**Pull request merged**
